Contract Kit 2.0.0

aaronmgdr opened the pull request:
# Description
 
 Major Version change to SDK packages. 
 
 ## Goals
 
1. Bundle Size of the flagship [@celo/contractkit package is as of latest version 3.2MB minified](https://bundlephobia.com/package/@celo/contractkit@1.5.2) Considering we are advertising this as being mobile that size is far to big. (15 seconds to download on 3g network) 

2. Minimize Breaking Changes 
This is not a complete rewrite. I want to make changes only where it made a big different to bundle size and the migration would not be too difficult 

### Result 

As of Alpha 4 Bundlephobia lists the bundle size as 1.7MB minified (which is still very big) Most of this comes from A) our contract ABIs/wrappers and B) Web3 itself. Therefore I would say we have reached the limit of how small this will get without drastic changes. However for Dapps needing a subset of contractkit there are alternatives `MiniContractKit` and `Connect`


## How We Got there

Prime source of  bloat was due to external packages imported via @celo/utils. To fix I split off 2 packages 

### @celo/phone-utils 

Functions which depend on either  `google-libphonenumber` or `country-data`

### @celo/cryptographic-utils 

Functions which depend on `bigi`, `bip39` `bip32`, `bls12377js`  `buffer-reverse`
 * from './account'
 * from './bls'
 * from './commentEncryption'
 * from './dataEncryptionKey'

 ### Refactoring
 
 Third pillar was refactoring of `WrapperCache` `AddressRegistry`  `CeloTokens`, and `**Wrapper`s so that they did not need a full `ContractKit` as a Parameter.  This means it is now possible to use just the parts of the Kit that one needs. For how this lead to breaking changes see that section.
 
 #### MiniKit

Furthermore it enabled the Creation of `MiniContractKit` which is a Kit that only provides access to `AccountsWrapper` The Token Wrappers, The Exchange Wrappers, and `GasPriceMinimumWrapper`. It gives no access to nor uses the `Web3ContractsCache` 

use Like 
```
import {ContractKit, newKitFromWeb3} from "@celo/contractkit/lib/mini-kit"


const miniKit = newKitFromWeb3(web3)

miniKit.setFeeCurrency()

miniKit.fillGasPrice()

/// full access to `connection` eg
 miniKit.connection.addAccount(privateKey)
 miniKit.connection.isSyncing()
miniKit.connection.sendTransaction(tx)

```
MiniKit has no Access to 
* Epoch/Block number functions (Now available on `ValidatorsWrapper`
* _web3_contracts_cache
* networkConfig


#### CeloTokens#isStableTokenContract 
This is now exported as a standalone function (kept on the instance for backwards compatibility)

#### ValidatorsWrapper 

Functions that used the validator wrapper but lived on the ContractKit have been moved here (with forwarding to kit for backwards compatibility)

```typescript

getEpochSize(): Promise<number> 

getFirstBlockNumberForEpoch(epochNumber: number): Promise<number> 
 
getLastBlockNumberForEpoch(epochNumber: number): Promise<number> 
 
getEpochNumberOfBlock(blockNumber: number): Promise<number> 
```

### @celo/base 
 enum for` StableTokens` `Token` `CeloTokens` move there from @celo/contractkit/celo-tokens (reexports for backwards compatibility


### @celo/connect 
* removed web3-utils ( some interfaces from there have been copied to abi-types.ts file )

# Other changes
* ALL Packages marked as `sideEffects: false`
* removed lodash from @celo/utils. as far as i could tell it was not used
* added "exclude": ["**/*.test.ts"] to tsconfig for the @celo/utils and phone-utils packages because before the lib folder had all our tests in it which was then published to NPM. which i think would increase the package size.

# Tested

# Related issues
Fixes Break up Utils Package #6189

# Backwards compatibility

## Breaking Changres

### @celo/contractkit

Most changes here are about eliminating the need to construct an entire kit to use classes and functions. and thus giving us the dream of a modular sdk


#### AddressRegistry

now takes an `Connection` instance instead of a `kit`

#### CeloTokens

no longer requires kit instead requires a Class implementing `ContractCacheType` to be passed. Examples are `WrapperCache` or `CeloTokensCache`


#### Wrappers 

Rather than take the full Kit Wrappers now construct like
```javascript
///Most Common
constructor(connection: Connection, contract: Contract)
// The Voting Contracts (Governance, Election, Validator, LockedGold, Slashers, and Attestations
constructor(connection: Connection, contract: Contract, wrapperCache: WrapperCache)  
 // Sorted Oracles
constructor(connection: Connection, contract: Contract, addressRegistry:  AddressRegistry)
```
The `WrapperCache` takes care of this while constructing them and most likely there will not be many situations where Wrappers were constructed  directly given they needed a kit before. 


##### AccountsWrapper 

`authorizeValidatorSigner` method now requires a `ValidatorsWrapper` be passed in as final argument. 

#### Web3ContractCache 
Instead of a `kit` requires only a `AddressRegistry` (uses AddressRegistry's web3 instances)

### @celo/utils

Anyone who was importing {PhoneNumberUtils, getCountryEmoji, getPhoneHash, LocalizedCountry, Countries, validatePhone, validateInput } from "@celo/utils" must now import from @celo/phone-utils

Furthermore the io.ts file was split up and E164PhoneNumberType, AttestationServiceStatusResponseType, AttestationServiceTestRequestType, AttestationRequestType, GetAttestationRequestType, AttestationRequest, AttestationResponseType, AttestationResponse must be imported from @celo/phone-utils too.

This is because all those types indirectly depend on E164PhoneNumberType which uses isE164NumberStrict. One alternative is to change this to use isE164Number from celo/base which is regeex based instead but im not sure of the affects of that. but it would greatly simplify a lot here.

# Documentation

Still in progress I Am adding readmes to all subpackages, making sure functions have typedoc annotations and new packages have typedoc setup. 

I plan to write a blog post / migration guide on docs.celo.org 

My goal is to have this all done before celo connect. and have beta 1 released before or durin.

aaronmgdr committed:
remove lodash from celo utils as it was unused

aaronmgdr committed:
Add Celo/phone-utils package

Everything from celo/utils which used google-libphonenumber or country.js has been move here include the io.ts file.

aaronmgdr committed:
Remove PhoneUtils, Country Functions and io from celo/utils

aaronmgdr committed:
update changelog

aaronmgdr committed:
mark package as being from of side effects for better treeshaking

aaronmgdr committed:
update deps graph

aaronmgdr committed:
temporarily reference phone utils as utils in faucet until the package is published since faucet must depend only on published packages

aaronmgdr committed:
run yarn.lock

aaronmgdr committed:
update deps grahp

aaronmgdr committed:
missed adding utils as dep of phone-utils

dckesler committed:
Adds deprecation script, upgrades sdks to 1.4.1

dckesler committed:
Bump WCv2-beta version, bump sdks to 1.4.2

dckesler committed:
Update sdks and cli to version 1.5.0

aaronmgdr committed:
Merge branch 'dckesler/sdks_cli-1.5.0' into aaronmgder/utils-breakup

# Conflicts:
#	packages/attestation-service/package.json
#	packages/faucet/package.json
#	packages/sdk/CHANGELOG.md
#	packages/sdk/identity/package.json
#	packages/sdk/phone-utils/package.json
#	packages/sdk/utils/package.json
#	packages/sdk/utils/src/io.ts
#	yarn.lock

aaronmgdr committed:
add phone-utils as a dev dep.

aaronmgdr committed:
update deps graph

aaronmgdr committed:
update to make it clear this is a breaking change

aaronmgdr committed:
see if this fixes the cant find 1.5.1-dev error

aaronmgdr committed:
ok so order is not the thing

aaronmgdr committed:
@celo/phone-ut

aaronmgdr committed:
untangle contractKit intra package dependencies.
Make it so nothing inside of the contractKit packgaes needs a kit to be passed into its constructor.

instead use Connection or WrapperCache or move functions close to their source.

This should alllow us to be more selective and decide to one ContractWrapper without needing to initialize a kit which itself would require initializing all bunch of other data.

This should lead to the ability to only import parts of contract kit. and lead to small bundle sizes

aaronmgdr committed:
fix types and rename class fix missed updates to required function parameters

aaronmgdr committed:
fix a test for Web3 cache which i broke by testing the ContractCache

aaronmgdr committed:
Import Code for Wrappers Async

aaronmgdr committed:
alpha2

aaronmgdr committed:
create cryptographic-utils package

aaronmgdr committed:
move AttestationServiceStatusResponseType back to utils

aaronmgdr committed:
fix packages home and repo and add some type docs + readme to connection

aaronmgdr committed:
fix dep graph and remove old cli command

nicolasbrugneaux committed:
chore: add prerelease checks to the deploy script

aaronmgdr committed:
alpha 3

aaronmgdr committed:
include phone utils in metadata crawler

aaronmgdr committed:
fix missing phone-utils

aaronmgdr committed:
get it working

aaronmgdr committed:
git rerun yarn

aaronmgdr committed:
Add cryptographic-utils where it is needed

aaronmgdr committed:
update deps graph

aaronmgdr committed:
updated imports

aaronmgdr committed:
lint cleanup

aaronmgdr committed:
fix import order for lint.

note that the linter wants the CAPITALIZED to go first, whereas the autosorter in vs code wants underscase first

aaronmgdr committed:
Merge branch 'master' into aaronmgder/ck-2

aaronmgdr committed:
simplify what gets passed, fix some tests, some still broken

aaronmgdr committed:
not gonna use these

aaronmgdr committed:
Add back the jest spy  😊 so the tests pass like they did before.

aaronmgdr committed:
Merge branch 'nicolasbrugneaux/safe-publish' into aaronmgder/ck-2

aaronmgdr committed:
update to latest BLS 12377

aaronmgdr committed:
update ts

aaronmgdr committed:
if fail remind to use npm

aaronmgdr committed:
alpha 4

aaronmgdr committed:
oh the yarn

aaronmgdr committed:
Merge branch 'master' into aaronmgder/ck-2

# Conflicts:
#	packages/protocol/package.json
#	yarn.lock

aaronmgdr commented on packages/sdk/contractkit/src/wrappers/Accounts.ts:
```diff
@@ -468,4 +465,10 @@ export class AccountsWrapper extends BaseWrapper<Accounts> {
   }
+
+  private keccak256(value: string | BN): string {
+    return this.connection.web3.utils.keccak256(value)
```
rather than import use the function on the web3 instance passed in

gastonponti replied:
If you move this function to the connection package, you can use `this.connection.keccak256(value)`, which will also clean even more the kit from web3.
This way web3 will only live in the connection package, and if we want to use ethers instead, we will have to only change those implementations (or wrappers in this case)

aaronmgdr commented on packages/sdk/contractkit/src/wrappers/Attestations.ts:
```diff
@@ -249,8 +267,11 @@ export class AttestationsWrapper extends BaseWrapper<Attestations> {
-    const tokenAddress = await this.kit.registry.addressFor(CeloContract.StableToken)
-    const attestationFee = await this.contract.methods.getAttestationRequestFee(tokenAddress).call()
+    const contract = await this.contracts.getStableToken(StableToken.cUSD)
+    // TODO confirm that contract.address === registry.addressFor(StableToken)
```
note to self

aaronmgdr commented on packages/sdk/cryptographic-utils/package.json:
```diff
@@ -0,0 +1,43 @@
+    "clean": "tsc -b . --clean",
+    "test": "jest --runInBand --ci",
+    "test:verbose": "jest --verbose",
+    "lint": "tslint -c tslint.json --project ."
```
todo add typedoc

aaronmgdr commented on packages/sdk/phone-utils/src/io.ts:
```diff
@@ -14,23 +14,6 @@ export const E164PhoneNumberType = new t.Type<string, string, unknown>(
     ),
   String
 )
-export const AttestationServiceStatusResponseType = t.type({
```
moved to Utils, so that AttewstationsWrapper does not need to import phone-utils

aaronmgdr commented on packages/sdk/phone-utils/typedoc.json:
```diff
@@ -0,0 +1,13 @@
+    "ignoreCompilerErrors": true,
+    "hideGenerator": "true",
+    "theme": "gitbook",
+    "out": "../../docs/developer-resources/utils/reference",
```
hmm should change this to phone-utils

aaronmgdr committed:
cleanup

aaronmgdr committed:
remove async contract loading as since we build to es5 it didnt help anything

aaronmgdr committed:
Add MiniKit: A minimal Version of Contract Kit with a subset but compatible interface.

remove old TODOs

Tests for miniKit

Move StableToken enum to Base

aaronmgdr committed:
Fix missed exports

aaronmgdr committed:
alpha 5

aaronmgdr committed:
lockfile

aaronmgdr committed:
probably need this one too

aaronmgdr committed:
Merge branch 'master' into aaronmgder/ck-2

aaronmgdr committed:
this should do it

aaronmgdr committed:
fix missing reference in docker

aaronmgdr committed:
these dockers need crypto utils

aaronmgdr committed:
fix missed missing cryptographic-utils ref

aaronmgdr commented on packages/cli/src/commands/account/authorize.ts:
```diff
@@ -83,7 +83,8 @@ export default class Authorize extends BaseCommand {
     } else if (res.flags.role === 'validator') {
-      tx = await accounts.authorizeValidatorSigner(res.flags.signer, sig)
+      const validatorsWrapper = await this.kit.contracts.getValidators()
+      tx = await accounts.authorizeValidatorSigner(res.flags.signer, sig, validatorsWrapper)
```
Passing in Validator New Required so that we dont need to have wrapperCache bring it in when constructing. this allows for the creation of MiniKit

aaronmgdr commented on packages/cli/src/commands/account/claims.test.ts:
```diff
@@ -28,8 +28,8 @@ testWithGanache('account metadata cmds', (web3: Web3) => {
     }
 
-    const readFile = () => {
-      return IdentityMetadataWrapper.fromFile(kit, emptyFilePath)
```
technically fromFile was async anyawy so `readFile` would have needed to be `await`ed no mater what

aaronmgdr commented on packages/sdk/connect/readme.md:
```diff
@@ -0,0 +1,13 @@
+const web3 = new Web3("YOUR_RPC_URL")
+const connection = new Connection(web3)
+
+const connectedChainID =  await connection.chainId()
```
Start of Readme for Connect package.

aaronmgdr replied:
PS Ive moved READme updates to a separate Branch as this one is A bit large already

aaronmgdr commented on packages/sdk/contractkit/src/kit.test.ts:
```diff
@@ -41,64 +42,66 @@ export function txoStub<T>(): TransactionObjectStub<T> {
-    const txRes = await kit.connection.sendTransactionObject(txo)
+    test('should send transaction on simple case', async () => {
+      const txo = txoStub()
+      txo.estimateGasMock.mockResolvedValue(1000)
+      const txRes = await kit.connection.sendTransactionObject(txo)
 
-    txo.resolveHash('HASH')
-    txo.resolveReceipt('Receipt' as any)
+      txo.resolveHash('HASH')
+      txo.resolveReceipt('Receipt' as any)
 
-    await expect(txRes.getHash()).resolves.toBe('HASH')
-    await expect(txRes.waitReceipt()).resolves.toBe('Receipt')
-  })
+      await expect(txRes.getHash()).resolves.toBe('HASH')
+      await expect(txRes.waitReceipt()).resolves.toBe('Receipt')
+    })
 
-  test('should not estimateGas if gas is provided', async () => {
-    const txo = txoStub()
```
This file changes is all just indentation.

Using same tests on MiniKit and ContractKit

aaronmgdr commented on packages/sdk/contractkit/src/kit.ts:
```diff
@@ -42,14 +44,7 @@ export const API_KEY_HEADER_KEY = 'apiKey'
  * @optional options to pass to the Web3 HttpProvider constructor
  */
 export function newKit(url: string, wallet?: ReadOnlyWallet, options?: HttpProviderOptions) {
-  let web3: Web3
-  if (url.endsWith('.ipc')) {
-    web3 = new Web3(new Web3.providers.IpcProvider(url, net))
-  } else if (url.toLowerCase().startsWith('http')) {
-    web3 = new Web3(new Web3.providers.HttpProvider(url, options))
-  } else {
-    web3 = new Web3(url)
-  }
```
moved to a function to be shared with mini-kit

aaronmgdr commented on packages/sdk/contractkit/src/kit.ts:
```diff
@@ -215,40 +203,22 @@ export class ContractKit {
+    return validators.getEpochSizeNumber()
   }
 
   async getFirstBlockNumberForEpoch(epochNumber: number): Promise<number> {
```
implementation moved to ValidatorsWrapper

gastonponti replied:
I think the epochSize is part of the majority of the contracts. Maybe is time to add that function to the wrapper of the blockchainParameters, which it seems more appropriate for that function (but this is a nice to have)

aaronmgdr replied:
I decided that I agreed that BlockchainParameters is indeed a better spot for this stuff and moved these group of functions over there plus added a test.

aaronmgdr commented on packages/sdk/contractkit/src/wrappers/Attestations.ts:
```diff
@@ -30,6 +29,25 @@ export function getSecurityCodePrefix(issuerAddress: Address) {
   return `${hashAddressToSingleDigit(issuerAddress)}`
 }
 
+// from '@celo/phone-utils/lib/io'
```
moved here to remove need for dependency on google lib phone number

aaronmgdr committed:
passing connection rather than web3 to Address Reg (make it easier to remove web3 in future)
Rather than require the entire WrapperCache to be passed into Wrappers, ask only for set of  those required.

aaronmgdr committed:
Merge branch 'master' into aaronmgder/ck-2

aaronmgdr committed:
fix missed param changes in tests setup

aaronmgdr committed:
Merge branch 'aaronmgder/ck-2' of https://github.com/celo-org/celo-monorepo into aaronmgder/ck-2

aaronmgdr committed:
Merge branch 'master' into aaronmgder/ck-2

# Conflicts:
#	packages/sdk/phone-utils/typedoc.json

gastonponti commented on packages/sdk/contractkit/src/address-registry.ts:
```diff
@@ -22,9 +22,9 @@ export class AddressRegistry {
+  constructor(readonly connection: Connection) {
     this.cache.set(CeloContract.Registry, REGISTRY_CONTRACT_ADDRESS)
-    this.registry = newRegistry(kit.connection.web3, REGISTRY_CONTRACT_ADDRESS)
+    this.registry = newRegistry(connection.web3, REGISTRY_CONTRACT_ADDRESS)
```
As we talked before, if we also change the way we generate the contracts, we can send the `connection` object instead the instance of `web3`. This will allow us to manage all our generated packages, without the need of web3, even if right now is under the hood, it will help us to change or swap that implementation without changing everything again.

if in this line we change the `newResgistry(web3, address)` to
`newRegistry(connection, address)`

and then in the generator of the file
https://github.com/celo-org/typechain-target-web3-v1-celo/blob/master/src/generation.ts#L38
we use something like
```
export function newRegistry(connection: Connection, address: string): Registry {
  return connection.contractMaker(ABI, address) as any
}
```

And then, at least for now, add in the connection object, something like
```
contractMaker(abi: AbiItem[], address: string) { \\Or a better name, haha
   return new this.web3.eth.Contract(abi, address)
}
```

It will allow us for example, to have a distro that does the same but with ethersjs, or even our own implementation of the Contracts.
It's not strictly necessary right now, to be honest, but we can use this mayor change version to address this.

aaronmgdr replied:
I do think this is a great idea, yet given that It would first involve updating the we3 celo typescript generator, and given how it by itself doesnt give us much, and given that the amount that has already been changed in v2 I think Its best to wait. 

I believe we could do this in steps by 

1. updating the generator template 
2. adding a `contractMaker` or `newContract` function to Connection
3. moving the generated contracts (sans web3 param) to their own package (would also make it possible to move sdk out of monorepo) 
4. to avoid breaking changes the contractkit could continue to reexport our the `new*` functions like exist now

gastonponti commented on packages/sdk/contractkit/src/contract-cache.test.ts:
```diff
@@ -14,10 +16,13 @@ const TestedWrappers: ValidWrappers[] = [
+  const web3 = new Web3('http://localhost:8545')
+  const connection = new Connection(web3)
+  const registry = new AddressRegistry(connection)
+  const web3ContractCache = new Web3ContractCache(registry)
```
In the same line of removing web3 from everywhere, we could rename the web3ContractCache, to something like, `coreContractCache`

aaronmgdr replied:
core sounds more like a subset of contracts to me. 

I could see this being contract cache and the existing being wrapper cache?

or to avoid reusing names `GeneratedContractsCache`


allthough it seems less important to do this at this time

gastonponti commented on packages/sdk/contractkit/src/contract-cache.ts:
```diff
@@ -44,141 +42,127 @@ const WrapperFactories = {
-  getDoubleSigningSlasher() {
-    return this.getContract(CeloContract.DoubleSigningSlasher)
+  getDoubleSigningSlasher(): Promise<DoubleSigningSlasherWrapper> {
+    return this.getContract<CeloContract.DoubleSigningSlasher>(CeloContract.DoubleSigningSlasher)
```
Is this cast necessary?

aaronmgdr replied:
was but no longer. 

i have undone

gastonponti commented on packages/sdk/contractkit/src/contract-cache.ts:
```diff
@@ -187,15 +171,25 @@ export class WrapperCache {
-      const instance = await this.kit._web3Contracts.getContract(contract, address)
-      const Klass: CFType[C] = WrapperFactories[contract]
-      this.wrapperCache[contract] = new Klass(this.kit, instance as any) as any
+      const instance = await this._web3Contracts.getContract<C>(contract, address)
```
Same with the `_web3Contracts` name

gastonponti commented on packages/sdk/contractkit/src/contract-cache.ts:
```diff
@@ -187,15 +171,25 @@ export class WrapperCache {
-  public invalidateContract<C extends ValidWrappers>(contract: C) {
-    this.kit._web3Contracts.invalidateContract(contract)
-    ;(this.wrapperCache[contract] as any) = null
+  public invalidateContract<C extends CeloContract>(contract: C) {
```
Shouldn't be a `ValidWrappers`?

gastonponti commented on packages/sdk/contractkit/src/identity/claims/verify.ts:
```diff
@@ -44,7 +44,11 @@ export const verifyAccountClaim = async (
   try {
-    metadata = await IdentityMetadataWrapper.fetchFromURL(kit, metadataURL, tries)
+    metadata = await IdentityMetadataWrapper.fetchFromURL(
+      await kit.contracts.getAccounts(),
```
For what I'm seeing, the kit is only used for the `kit.contracts.getAccounts()`
Knowing that we are already making a major change, why don't we also change this to be able to use the mini-kit too?

```
export async function verifyClaim(accounts: AccountsWrapper, claim: Claim, address: string, tries = 3) {
  switch (claim.type) {
    case ClaimTypes.KEYBASE:
      return verifyKeybaseClaim(accounts, claim, address)
    case ClaimTypes.ACCOUNT:
      return verifyAccountClaim(accounts, claim, address, tries)
    case ClaimTypes.DOMAIN:
      return verifyDomainRecord(accounts, claim, address)
    default:
      break
  }
  return
}
```

And change every claim to use the accounts instead

gastonponti commented on packages/sdk/contractkit/src/identity/claims/keybase.ts:
```diff
@@ -34,7 +34,7 @@ export async function verifyKeybaseClaim(
 
     const hasValidSignature = await IdentityMetadataWrapper.verifySignerForAddress(
-      kit,
+      await kit.contracts.getAccounts(),
```
same, send the accounts instead the kit

gastonponti commented on packages/sdk/contractkit/src/mini-kit.ts:
```diff
@@ -0,0 +1,107 @@
+    this.connection.defaultFeeCurrency = address
+  }
+
+  // TODO: remove once cUSD gasPrice is available on minimumClientVersion node rpc
```
This is not required anymore. The gas estimation is done in the client since the 1.3.x version, as we are on the 1.5.x which was a hardfork, all our nodes should have that in the client

gastonponti commented on packages/sdk/contractkit/src/mini-kit.ts:
```diff
@@ -0,0 +1,107 @@
+    await this.connection.setGasPriceForCurrency(currency, gasPrice)
+  }
+
+  async fillGasPrice(tx: CeloTx): Promise<CeloTx> {
```
Part of the same that could be removed

aaronmgdr committed:
Merge branch 'master' into aaronmgder/ck-2

gastonponti commented on packages/sdk/contractkit/src/mini-kit.ts:
```diff
@@ -0,0 +1,107 @@
+      tokenContract === CeloContract.GoldToken
+        ? undefined
+        : await this.registry.addressFor(tokenContract)
+    if (address) {
```
Not necessary anymore

gastonponti commented on packages/sdk/contractkit/src/celo-tokens.ts:
```diff
@@ -1,27 +1,18 @@
 import { GoldTokenWrapper } from './wrappers/GoldTokenWrapper'
 import { StableTokenWrapper } from './wrappers/StableTokenWrapper'
-
-export enum StableToken {
```
I think this was moved to the /base package, right?
I think, that probably that will require more packages to be published in the case of adding a new stable.
Don't really know if it was required in the base for something in particular, but if this is not part of the CK, I think that should be part of the connection one, not necessarily the base one

crondinini replied:
What was the gain (or goal) from moving this to another package?

gastonponti commented on packages/sdk/contractkit/src/web3-contract-cache.ts:
```diff
@@ -82,8 +94,8 @@ type ContractCacheMap = { [K in keyof CFType]?: ReturnType<CFType[K]> }
  */
 export class Web3ContractCache {
```
As I told you before, maybe it's a good time to change this name (if we want also to remove web3 from the generated packages)

gastonponti commented on packages/sdk/contractkit/src/web3-contract-cache.ts:
```diff
@@ -173,11 +185,14 @@ export class Web3ContractCache {
       debug('Initiating contract %s', contract)
       const createFn = ProxyContracts.includes(contract) ? newProxy : ContractFactories[contract]
-      this.cacheMap[contract] = createFn(this.kit.connection.web3, address) as ContractCacheMap[C]
+      this.cacheMap[contract] = createFn(
```
If we change the way we create the contracts (adding that wrapper to the connection package), we can also refactor this part

aaronmgdr replied:
true

gastonponti commented on packages/sdk/contractkit/src/wrappers/Election.ts:
```diff
@@ -482,6 +485,10 @@ export class ElectionWrapper extends BaseWrapper<Election> {
   }
 
+  private validators() {
+    return this.contracts.getValidators()
```
if every time you are calling this one, you are using await, why not using it here? (and keep the other thing cleaner)

aaronmgdr replied:
( im  pretty sure) when returning a promise it doesn't really do anything to await. 

But this private function isnt that needed so im just gonna remove

gastonponti commented on packages/sdk/contractkit/src/wrappers/Reserve.ts:
```diff
@@ -71,7 +71,7 @@ export class ReserveWrapper extends BaseWrapper<Reserve> {
     this.contract.methods.getAssetAllocationSymbols,
     undefined,
-    (symbols) => symbols.map(this.kit.web3.utils.hexToAscii)
+    (symbols) => symbols.map(this.connection.web3.utils.hexToAscii)
```
again, maybe extract this to the connection package, to be able to easily remove it in the future

gastonponti commented on packages/sdk/utils/src/address.ts:
```diff
@@ -43,4 +41,20 @@ export const publicKeyToAddress = (publicKey: string) =>
   privateKey.startsWith('0x') && isValidPrivate(hexToBuffer(privateKey))
 
-export const isValidAddress = (input: string): boolean => Web3Utils.isAddress(input)
+export const isValidAddress = (input: string): boolean => {
```
❤️

aaronmgdr committed:
keep interface like full kit

aaronmgdr committed:
review from Ponti

-- add wrapper functions for some used web3 utils to connection to make it easier to eventually remove web3

-- fix type on invalidateContract

-- remove unneeded function

-- move functions down in file since they are in the way of another function and its typedoc

-- add getEpochNumber and get EpochSize to blockchain parameters

aaronmgdr committed:
PR Review from Ponti

- calls to get contract don't need to be explicitly type
- move block Number and epoch number functions to blockchain params
- fix name of LockedGold test
- add Test for Blockchainparameters wrapper
- fix hexToAscii not receiving `this` when called from Reserve Wrapper

aaronmgdr committed:
Merge branch 'master' into aaronmgder/ck-2

aaronmgdr committed:
use the same node types through the monorepo

aaronmgdr committed:
Merge branch 'aaronmgder/ck-2' of https://github.com/celo-org/celo-monorepo into aaronmgder/ck-2

crondinini commented on packages/sdk/contractkit/src/identity/metadata.ts:
```diff
@@ -38,39 +38,38 @@ export class IdentityMetadataWrapper {
 
-  static fromFile(kit: ContractKit, path: string) {
-    return this.fromRawString(kit, readFileSync(path, 'utf-8'))
+  static fromFile(accounts: AccountsWrapper, path: string) {
```
I'm wondering if you explored the possibility of having a proxy function that accepts both `ContractKit` and `AccountsWrapper`. I don't know how much this package is used externally or in which context (beyond the changes in this PR), but it seems a big breaking change.

crondinini replied:
You mentioned in your PR _"This change was done so that kit was not required to be passed into all the classes and functions that use IdentityMetadataWrapper"_. This makes sense, but I'd love to understand why? In the CLI changes for example, you are using kit to the accounts, so there's not much gain in that specific case, so I'd like to understand your vision for this change.

aaronmgdr replied:
The idea is that when you require `kit` as the argument, you are requiring the consumer to have a class that brings with it the code for every contract wrapper and contract the kit gives access to. That's a lot of code and bytes where all thats really required is one function on one Wrapper. 

While it is true that the CLI already has a kit and will still use one. in the AttestationsWrapper  (which uses this IdentityMetaData functions) now all thats is required is one or two contracts to initialize an instance

aaronmgdr replied:
I do like the idea of allowing it to take either and thus not have a breaking change

crondinini commented on packages/docs/developer-resources/utils/reference/README.md:
```diff
@@ -0,0 +1,162 @@
+  </a>
+</p>
+
+**Celo Monorepo - Official repository for core projects comprising the Celo platform**
```
🤔 Is this README correct? Seems to be talking about the monorepo and not utils. Also, where will it be displayed?

aaronmgdr replied:
Im gonna need to take a look at the whole `/docs` part of this

crondinini commented on packages/sdk/CHANGELOG.md:
```diff
@@ -1,10 +1,21 @@
+
+
+  - @celo/utils
+      * comment encryption, bls and mneumonic functions moved to @celo/cryptographic-utils
```
Shouldn't we add also all other  signatures that changed? i.e.: some methods in `IdentityWrapper` plus others that I don't remember off the top of my head.

crondinini commented on packages/sdk/CHANGELOG.md:
```diff
@@ -1,10 +1,21 @@
+
+### **[2.0.0-dev]**
+Breaking Change
+  - Removes phone and country related functions from utils. Now in phone-utils [#](https://github.com/celo-org/celo-monorepo/pull/8987)
```
Would be interesting to specify the functions here to avoid the trouble of the person having to go to this PR and having to read through everything.

crondinini commented on packages/sdk/connect/src/connection.ts:
```diff
@@ -38,13 +38,20 @@ import { ReadOnlyWallet } from './wallet'
 }
 
+/**
+ * Connection a Class for connecting to Celo, sending Transactions, etc
```
```suggestion
 * Connection is a Class for connecting to Celo, sending Transactions, etc
```

crondinini commented on packages/sdk/connect/src/connection.ts:
```diff
@@ -86,6 +93,16 @@ export class Connection {
     }
   }
 
+  // functions which for now call web3 functions but may do something else later
```
Is this comment needed? If it's for the dev importing this function, should we add it to "doc"/comment of the function itself?

crondinini commented on packages/sdk/contractkit/src/celo-tokens.ts:
```diff
@@ -68,7 +59,7 @@ export const celoTokenInfos: {
  */
 export class CeloTokens {
-  constructor(readonly kit: ContractKit) {}
+  constructor(readonly contracts: ContractCacheType, readonly registry: AddressRegistry) {}
```
Also wonder if it'd be useful to make this backwards compatible?

aaronmgdr replied:
I think that would be of limited use. In the past this required a ContractKit Instance, but creating an instance of ContractKit itself makes an instance of CeloTokens so in the past it would have been very odd to initialize this directly

crondinini replied:
Makes sense!

crondinini commented on packages/sdk/cryptographic-utils/src/bls.ts:
```diff
@@ -1,9 +1,9 @@
-const keccak256 = require('keccak256')
+import { isValidAddress } from '@celo/utils/lib/address'
+import { BLS } from 'bls12377js'
+import { keccak256 } from 'ethereumjs-util'
```
Why replace the `keccak256` pkg for `ethereumjs-util`?

aaronmgdr replied:
this was done to reduce the number of package deps

crondinini commented on packages/sdk/utils/src/io.ts:
```diff
@@ -98,98 +81,3 @@ export const AttestationServiceStatusResponseType = t.type({
   maxRerequestMins: t.number,
   twilioVerifySidProvided: t.boolean,
 })
```
Does the comment apply to this too? Or do we still need `AttestationServiceStatusResponseType`?

aaronmgdr replied:
At one point I thought moving AttestationServiceStatusResponseType would be beneficial so i marked it as deprecated here. but it turns out I was wrong. So brought it back

aaronmgdr committed:
Update packages/sdk/connect/src/connection.ts

Co-authored-by: Camila Rondinini <crondinini@gmail.com>

aaronmgdr committed:
Dont include old gassPrice logic
since the min blockchain node version now all return a gasPrice for each fee payable currency,
these functions to calculate gasPrice on the dapp/wallet side are no longer needed as such they will not be included in minikit

GasPriceMinimum Contract is not required

aaronmgdr committed:
Merge branch 'aaronmgder/ck-2' of https://github.com/celo-org/celo-monorepo into aaronmgder/ck-2

aaronmgdr committed:
make IdentityMapper accepting AccountsWrapper  non breaking change

aaronmgdr committed:
move this test as minitest no longer supports this deprecated gas method

aaronmgdr committed:
2.0.0 Beta 1

aaronmgdr committed:
forgot that yarn lock!

aaronmgdr committed:
Merge branch 'master' into aaronmgder/ck-2

# Conflicts:
#	packages/phone-number-privacy/common/package.json
#	packages/sdk/CHANGELOG.md
#	packages/sdk/encrypted-backup/package.json

aaronmgdr committed:
remove comment

aaronmgdr committed:
move to version of bls12377 which uses ts 4

aaronmgdr committed:
fix lockfile

aaronmgdr committed:
Beta 2

aaronmgdr committed:
update bls

aaronmgdr committed:
yarn file

aaronmgdr committed:
update change log

aaronmgdr committed:
Merge branch 'master' into aaronmgder/ck-2

aaronmgdr committed:
SDK READMES & TypeDocs (#9418)

* Add Basic Readme files to the sdk packages

* WIP readmes

* add  more  readmes, provide descriptions of wallet packages

* Set packages homepage to the sdk docs site

* mark functions as deprecated or internal

* Add details to typdoc comments,
set readmes for a few packages

* mark functions as internal and other cleanup

* lint

* clean up docs

* edits

* formatting

* formatting

* link to example repo

* improve readme for kit

* Update packages/sdk/connect/readme.md

Co-authored-by: Camila Rondinini <crondinini@gmail.com>

* Update packages/sdk/connect/readme.md

Co-authored-by: Camila Rondinini <crondinini@gmail.com>

* its a wrapper not technically a contract

* Update packages/sdk/contractkit/README.md

Co-authored-by: Camila Rondinini <crondinini@gmail.com>

* Add reference to upgrade Guide in Changelog

* Update packages/sdk/connect/src/utils/abi-utils.ts

Co-authored-by: Camila Rondinini <crondinini@gmail.com>

* Fix missing contract in readme example

* generate docs

Co-authored-by: Josh <critesjosh@gmail.com>
Co-authored-by: Camila Rondinini <crondinini@gmail.com>

aaronmgdr committed:
Merge branch 'master' into aaronmgder/ck-2

aaronmgdr committed:
2.0 Release Candidate

aaronmgdr committed:
RC 2.

dckesler committed:
Fix CLI build issues

aaronmgdr committed:
bind the correct this

aaronmgdr committed:
shrinkwrap

aaronmgdr committed:
CLI RC1

aaronmgdr committed:
shrink change

aaronmgdr committed:
Merge branch 'master' into aaronmgder/ck-2

aaronmgdr committed:
move back to -dev

aaronmgdr committed:
lint import order

aaronmgdr committed:
yes this is duplicated but its very small

aaronmgdr committed:
ok then

aaronmgdr committed:
Merge branch 'master' into aaronmgder/ck-2

aaronmgdr committed:
Merge branch 'master' into aaronmgder/ck-2

aaronmgdr committed:
RC 3 for good measure

aaronmgdr committed:
yarn again

aaronmgdr committed:
SDKs 2.0

aaronmgdr committed:
?
